File: mappers-lua/202.lua
--
-- http://wiki.nesdev.com/w/index.php/INES_Mapper_202
-- 

MapperName = "Mapper 202"
MapperNumber = 202
DefaultPrgSize = 8*16*1024
DefaultChrSize = 8*8*1024

function DumpPrg(size)
	local banks = math.floor(size / 0x4000)
	for b = 0, banks-1 do		
		print("Reading PRG bank #" .. tostring(b) .. "...")
		WriteCpu(0x8000 + b * 2, {0})
		ReadAddPrg(0xC000, 0x4000)
	end
end

function DumpChr(size)
	local banks = math.floor(size / 0x2000)
	for b = 0, banks-1 do		
		print("Reading CHR bank #" .. tostring(b) .. "... ")
		WriteCpu(0x8000 + b * 2, {0})
		ReadAddChr(0x0000, 0x2000)
	end
end

function EnablePrgRam(size)
	print("Warning: SRAM is not supported by this mapper")
end

File: FamicomDumper/mappers/202.cs
class Mapper202 : IMapper
{
    public string Name { get => "Mapper 202"; }
    public int Number { get => 202; }
    public int DefaultPrgSize { get => 8 * 0x4000; }
    public int DefaultChrSize { get => 8 * 0x2000; }

    public void DumpPrg(IFamicomDumperConnection dumper, List<byte> data, int size)
    {
        var banks = size / 0x4000;
        for (var bank = 0; bank < banks; bank++)
        {
            Console.Write($"Reading PRG bank #{bank}/{banks}... ");
            dumper.WriteCpu((ushort)(0x8000 | (bank << 1)), 0);
            data.AddRange(dumper.ReadCpu(0xC000, 0x4000));
            Console.WriteLine("OK");
        }
    }

    public void DumpChr(IFamicomDumperConnection dumper, List<byte> data, int size)
    {
        var banks = size / 0x2000;
        for (var bank = 0; bank < banks; bank++)
        {
            Console.Write($"Reading CHR bank #{bank}/{banks}... ");
            dumper.WriteCpu((ushort)(0x8000 | (bank << 1)), 0);
            data.AddRange(dumper.ReadPpu(0x0000, 0x2000));
            Console.WriteLine("OK");
        }
    }

    public MirroringType GetMirroring(IFamicomDumperConnection dumper)
    {
        return MirroringType.MapperControlled;
    }
}
